File: ReOrder List.py
#Time Complexity: O(n)
#Space Complexity: O(1)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
       
        mid=self.Mid(head);
        midnext=mid.next;
        mid.next=None;
        ReverseElementsAfterMid=self.reverse(midnext);
        root=head;
        return self.Merge(root,ReverseElementsAfterMid);
            
    def Merge(self,root,Reverse):
        list1=root;
        list2=Reverse;
        while list2:
            temp=list1.next
            list1.next=list2;
            list2=list2.next;
            list1.next.next=temp;
            list1=temp;
        return list1;
        
        
        
    def Mid(self,head):
        slow=head;
        fast=head;
        while fast.next and fast.next.next:
            slow=slow.next;
            fast=fast.next.next;
        return slow;
    
    def reverse(self,midnext):
        previous=None;
        current=midnext;
        while current:
            nxt=current.next;
            current.next=previous;
            previous=current;
            current=nxt;
        return previous;
        
            

File: ReorderList.java
/**
 * Time Complexity - O(n)
 * Space Complexity - O(1)
 */

class Solution {
    public void reorderList(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        while(fast != null && fast.next != null) {
            slow = slow.next;
            fast= fast.next.next;
        }

        ListNode midNext = slow.next;
        slow.next = null;
        ListNode previous = null;

        while(midNext != null) {
            ListNode temp = midNext.next;
            midNext.next = previous;
            previous = midNext;
            midNext = temp;
        }

        slow = head;
        midNext = previous;
        while(midNext != null) {
            ListNode slowNext = slow.next;
            ListNode midNodeNext = midNext.next;
            slow.next = midNext;
            midNext.next = slowNext;
            slow = slowNext;
            midNext = midNodeNext;
        }

    }
}